src/Elastic.Apm/Helpers/UrlUtils.cs

```diff
@@ -7,26 +8,23 @@ internal static class UrlUtils
 	{
 		private const string ThisClassName = nameof(UrlUtils);
 
-		internal static bool TryExtractDestinationInfo(Uri url, out string host, out int? port, IApmLogger logger)
+		/// <returns><c>Destination</c> if successful and <c>null</c> otherwise</returns>
+		internal static Destination TryExtractDestination(Uri url, IApmLogger logger)
```

Comment:
I know, we don't have an official convention about `TryParse` or `TryExtract` method signatures, however, for C# world is more convenient to return `bool` value and pass `out` parameter. Maybe it would be better to rename this method to not confuse people. WDYT?

Reply:
I'm not sure it's that clear cut - IMHO `Try` prefix means that error is signaled by using return value and not by using exceptions. Exceptions is much better error handling mechanism (at least IMHO - for example Rust and Go people disagree :) so the motivation for using a different (and worse) error handling mechanism is performance. Throwing an exception in .NET [has significant performance cost](https://mattwarren.org/2016/12/20/Why-Exceptions-should-be-Exceptional/) so it should be reserved for truly exceptional situations. Of course, sometimes it's not easy to decide if certain situations are "exceptional" or "expected". That is why some libraries provide both `TryXyz()` and `Xyz()` APIs if there's a chance that `Xyz()` will be used as part of hot code path.
So assuming we agree that `Try` prefix just means error via return value then it's just a question of whether return value type has a special value that is never returned in case of success. For APIs such as `int.TryParse()` or `Dictionary.TryGetValue()` there is no special value to signal error so you have to use `bool` to signal error and `out` for actual return value. But in `TryExtractDestination` case `null` cannot be returned in case of success so it can be used to signal error.
I do agree with you that it becomes a question of choosing between consistency with already existing APIs and making APIs more concise by using modern features such as non-nullable references (or maybe tuples). Regarding consistency: for example `System.Uri.TryCreate()` uses `bool/out` even though it could have returned `null` to signal error. But OTOH sometimes BCL sacrifices consistency to make API easier to use - for example `Enumerable.FirstOrDefault` returns null if the sequence is empty. Regarding making API easier to use - especially now with C# 8  non-nullable references having `SomeClass? TryXyz()` should make it quite clear that API can return `null` and most likely to signal a error. You can see discussions about it [here](https://softwareengineering.stackexchange.com/questions/387674/c-8-non-nullable-references-and-the-try-pattern) and [here](https://stackoverflow.com/questions/3619691/check-reference-parameter-value-or-return-bool).

In the end I decided to split the difference by removing `Try` prefix and keeping the doc comment clarifying that `null` is returned on error.

Reply:
Great explanation of your thoughts.

>In the end I decided to split the difference by removing Try prefix and keeping the doc comment clarifying that null is returned on error.

This is exactly what I meant, but maybe I shared it in a complicated way.